PXC-3567 Merge codership/mysql-wsrep 5.7.32 -> PXC 5.7

venkatesh-prasad-v opened the pull request:
https://jira.percona.com/browse/PXC-3567

Merge conflicts resolved

***
Updated the WSREP_PATCH_VERSION to 49.

***
Updated WSREP-REVISION to reflect the upstream refresh revision.
(commit 1e88d72).

***
Reverted upstream fix for
"Fix test of '-k' option for timeout(1) in wsrep_sst_xtrabackup-v2"
(commit 291cd9f and 1b073cd)
in favour of
"Fix error handling when donor is not able to serve SST"
https://jira.percona.com/browse/PXC-3424
(commit eddd130 and 647e728)

***
Corrected the fix for the bug
"Xid events for LOAD DATA splitted transactions"
(commit 660b5f4)
to include the check in both `wsrep_replicate()` and
`wsrep_run_wsrep_commit()` hooks and to fix the test failure of
galera.galera_load_data testcase.

***
Corrected the filtering logic in `check_table_map()` introduced by
"Replication filters should not affect galera cluster events"
(commit cf3987b)
to fix the test failure of galera.galera_repl_filtersA and
galera.galera_repl_filtersB test cases.

***
Adjusted the test galera.mysql-wsrep#372.test to

1. Replace occurences of `SET SESSION wsrep_on=ON/OFF` into
`SET GLOBAL wsrep_on=ON/OFF`, as `wsrep_on` is a session variable in PXC.

2. Add a suppression to the expected warnings.

***
Adjusted the test galera.galera_UK_conflict.test to use the correct
thread state `wsrep: replaying write set(..)`.

***
Re-recorded the following test cases.

1. main.mysqld--help-notwin
2. galera.galera_defaults
3. galera.galera_load_data
4. perfschema.show_sanity
5. suite.all_vars

Testing Done
---
**Round 1 (01-Feb-21)**
1. Galera and sysvar suite: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/101/testReport/
Parameters: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/101/parameters/
```
Failing Tests: 

1. galera.galera_sst_xtrabackup-v2_encrypt_with_key - Fails due to PXC-3508
2. galera.galera_sst_xtrabackup-v2_keyring - Fails due to PXC-3508
```

2. Full MTR: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/102/console
Parameters: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/102/parameters/
```
Failing Tests:

1. binlog.binlog_query_filter_rules - Fixed now
2. keyring_vault tests: https://jira.percona.com/browse/PXC-3297
3. main.merge_mmap & main.merge: https://jira.percona.com/browse/PXC-3296
4. main.commit_1innodb: Sporadic failure due to improper cleanup by its previous test main.bug93770. https://jira.percona.com/browse/PS-7479
5. galera.galera_sst_xtrabackup-v2_encrypt_with_key - Fails due to PXC-3508 
6. galera.galera_sst_xtrabackup-v2_keyring - Fails due to PXC-3508
```
**Round 2 (02-Feb-21)**
1. Full MTR: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/105/console
Parameters: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/105/parameters/
```
Failing Tests:
1. keyring_vault tests: https://jira.percona.com/browse/PXC-3297
2. main.merge_mmap & main.merge: https://jira.percona.com/browse/PXC-3296
3. main.commit_1innodb: Sporadic failure due to improper cleanup by its previous test main.bug93770. https://jira.percona.com/browse/PS-7479
4. galera.galera_sst_xtrabackup-v2_encrypt_with_key - Fails due to PXC-3508 
5. galera.galera_sst_xtrabackup-v2_keyring - Fails due to PXC-3508
```

bkandasa committed:
Update License Book

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

Commit:
codership/mysql-wsrep#372 Replication filters should not affect galera cluster events

Events originating from galera cluster should not be affected by replicate-do* filters.

Introduce new variable `wsrep_mode` which can be used to ignore native replication
filters for galera originating events.

Commit:
Bug #31472704 : MYSQL CRASHES ON SOME QUERIES WITH SUBSELECTS

Description:
Fix an issue with SELECT query.

Change-Id: I90a256bf237f1ce7860b1b3cfb25813287036607
(cherry picked from commit b3039dc34acc2544f5451b7d55ff239cc4c268a9)

temeo committed:
Merge tag 'mysql-5.6.50' into 5.6-v25-5.6.50

temeo committed:
Merge tag 'mysql-5.7.32' into 5.7-v25-5.7.32

temeo committed:
Null merge remote-tracking branch 'origin/5.6-v25' into 5.7-v25-null

sjaakola committed:
MDEV-24097 node restart overlaps with earlier still ongoing SST process

Merged the fix for lingerig SST rsync process preventing node startup, from MariaDB

sjaakola committed:
UK GAP lock conflict and replaying

MariaDB 10.5 got fix for MDEV-23851, which is about false positive lock conflict between appliers and local committer.
MySQL 5.7-v25 does not have similar issue, but the mtr test created for MDEV-23851 is backported in this commit.

Commit:
Restore wsrep recovery step in mysqld_safe

Due to a mistake during merging, mysqld_safe no longer performs
the recover step that is required to determine a proper starting
position for Galera.
This patch restores the wsrep recover step in mysqld_safe and adds a
MTR test that verifies that the recover step is performed.

Fixes codership/mysql-wsrep-bugs#833

Commit:
Fix test of '-k' option for timeout(1) in wsrep_sst_xtrabackup-v2

wsrep_sst_xtrabackup-v2 script tests if timeout command has option
`-k`. However this test fails, if `timeout --help` outputs to stderr.
The fix consists of redirecting the output of `timeout --help` from
stderr to stdout.

Fixes codership/mysql-wsrep#379

Commit:
Merge remote-tracking branch 'origin/5.6-v25-mysql-wsrep#379' into 5.7-v25-mysql-wsrep#379

ayurchen committed:
Fixing race in MTR test galera_ist_gcache_rollover

ayurchen committed:
Merge branch 'galera-bugs#592-5.6' into galera-bugs#592-5.7

sjaakola committed:
codership/mysql-wsrep#376 Xid events for LOAD DATA splitted transactions

Sub transactions for LOAD DATA splitting are not properly registered in mysql server side,
and the transaction commit and re-initialization happens only in innodb. This causes
transaction binlogging to skip creating XID event for the sub transactions.
Ths PR has additional logic in transaction binlogging phase to make sure that LOAD
DATA subtransactions will get the Xid event as well.

galera.galera_load_data test has new phase to verify that all LOAD DATA subtransactions
have Xid event

sjaakola committed:
codership/mysql-wsrep#376 Xid events for LOAD DATA splitted transactions

Added cleanup of temporary load data file

temeo committed:
Merge man pages for 5.6.50-25.32

temeo committed:
Bump wsrep patch version to 25.32

temeo committed:
Null merge branch '5.6-v25-5.6.50-25.32' into 5.7-v25-5.7.32-25.24

temeo committed:
Merge man pages for 5.7.32-25.24

temeo committed:
Bump wsrep patch version to 25.24

Commit:
Implemented PXC-3567 (Merge codership/mysql-wsrep 5.7.32 -> PXC 5.7)

https://jira.percona.com/browse/PXC-3567

Merge with conflicts.

Merge tag 'wsrep_5.7.32-25.24' into codership_5.7.32_merge

Conflicts:
    both modified:   cmake/wsrep.cmake
    both modified:   mysql-test/r/mysqld--help-notwin.result
    both modified:   mysql-test/suite/galera/r/galera_defaults.result
    both modified:   mysql-test/suite/galera/r/galera_load_data.result
    both modified:   mysql-test/suite/perfschema/r/show_sanity.result
    both modified:   scripts/mysqld_safe.sh
    both modified:   scripts/wsrep_sst_xtrabackup-v2.sh
    both modified:   sql/wsrep_mysqld.cc
    both modified:   sql/wsrep_mysqld.h
    both modified:   storage/innobase/handler/ha_innodb.cc

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/not_signed)](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1368) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1368) before we can accept your contribution.<br/>**1** out of **8** committers have signed the CLA.<br/><br/>:white_check_mark: venkatesh-prasad-v<br/>:x: bkandasa<br/>:x: gipulla<br/>:x: sjaakola<br/>:x: temeo<br/>:x: mkaruza<br/>:x: ayurchen<br/>:x: sciascid<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/percona-xtradb-cluster?pullRequest=1368) it.</sub>

kamil-holubicki commented on mysql-test/suite/galera/t/galera_UK_conflict.test:
```diff
@@ -91,7 +91,8 @@ SET GLOBAL debug = NULL;
 
 # wait for BF abort to happen and replaying begin
---let $wait_condition = SELECT COUNT(*)=1 FROM INFORMATION_SCHEMA.PROCESSLIST WHERE STATE= 'wsrep replaying trx' 
+
```
extra empty line

venkatesh-prasad-v replied:
fixed.

kamil-holubicki commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -9490,13 +9385,14 @@ ha_innobase::write_row(
 
-		switch (
-		    wsrep_run_wsrep_commit(m_user_thd, wsrep_hton, true)) {
+		int ret=0;
```
Wouldn't the following be more readable?
```
int ret = foo();
switch(ret) { ...
```

venkatesh-prasad-v replied:
Agreed. Done.

kamil-holubicki reviewed (commented):
Looks good. 
2 comments/questions

Commit:
Implemented PXC-3567 (Merge codership/mysql-wsrep 5.7.32 -> PXC 5.7)

https://jira.percona.com/browse/PXC-3567

Merge conflicts resolved

***
Updated the WSREP_PATCH_VERSION to 49.

***
Updated WSREP-REVISION to reflect the upstream refresh revision.
(commit 1e88d72).

***
Reverted upstream fix for
"Fix test of '-k' option for timeout(1) in wsrep_sst_xtrabackup-v2"
(commit 291cd9f4 and 1b073cde)
in favour of
"Fix error handling when donor is not able to serve SST"
https://jira.percona.com/browse/PXC-3424
(commit eddd1301 and 647e728c)

***
Corrected the fix for the bug
"Xid events for LOAD DATA splitted transactions"
(commit 660b5f48)
to include the check in both `wsrep_replicate()` and
`wsrep_run_wsrep_commit()` hooks and to fix the test failure of
galera.galera_load_data testcase.

***
Corrected the filtering logic in `check_table_map()` introduced by
"Replication filters should not affect galera cluster events"
(commit cf3987b4)
to fix the test failure of galera.galera_repl_filtersA and
galera.galera_repl_filtersB test cases.

***
Adjusted the test galera.mysql-wsrep#372.test to

1. Replace occurences of `SET SESSION wsrep_on=ON/OFF` into
`SET GLOBAL wsrep_on=ON/OFF`, as `wsrep_on` is a session variable in PXC.

2. Add a suppression to the expected warnings.

***
Adjusted the test galera.galera_UK_conflict.test to use the correct
thread state `wsrep: replaying write set(..)`.

***
Re-recorded the following test cases.

1. main.mysqld--help-notwin
2. galera.galera_defaults
3. galera.galera_load_data
4. perfschema.show_sanity
5. suite.all_vars

kamil-holubicki reviewed (approved):
lgtm
